jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/SparkParallelismTracker.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ Copyright (c) 2014 by Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.scala.spark
+
+import java.net.URL
+
+import ml.dmlc.xgboost4j.java.XGBoostError
+import org.apache.commons.logging.LogFactory
+import org.apache.spark.SparkContext
+import org.codehaus.jackson.map.ObjectMapper
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.{Failure, Success}
+import scala.collection.JavaConverters._
+
+/**
+  * A tracker that periodically checks the number of alive Spark executor cores.
+  * Throws an exception when the number of alive cores is less than nWorkers.
+  *
+  * @param sc The SparkContext object
+  * @param checkInterval The interval to check executor status in milliseconds.
+  * @param nWorkers nWorkers used in an XGBoost Job
+  */
+private[spark] class SparkParallelismTracker(
+    sc: SparkContext,
+    checkInterval: Long,
+    nWorkers: Int) {
+
+  private[this] val mapper = new ObjectMapper()
+  private[this] val logger = LogFactory.getLog("XGBoostSpark")
+
+
+  private[this] def isHealthy: Boolean = {
+    val numAliveCores = try {
+      val url = new URL(s"${sc.uiWebUrl.get}/api/v1/applications/${sc.applicationId}/executors")
+      mapper.readTree(url).findValues("totalCores").asScala.map(_.asInt).sum
+    } catch {
+      case ex: Throwable =>
+        logger.warn(s"Unable to read total number of alive cores from REST API." +
+          s"Health Check will be ignored.")
+        ex.printStackTrace()
+        Int.MaxValue
+    }
+    if (numAliveCores >= nWorkers) {
+      true
+    } else {
+      throw new XGBoostError(s"Requires numParallelism = $nWorkers but only " +
+        s"$numAliveCores cores are alive. Please check logs in Spark History Server.")
+    }
+  }
+
+  /**
+    * Execute a blocking function call with periodic checks on number of alive cores
+    *
+    * @param body A blocking function call
+    * @tparam T Return type
+    * @return The return of body
+    */
+  def execute[T](body: => T): T = {
```

Comment:
The logic here is quite counter-intuitive. You assume `body` is a blocking method? I need more eyes to check it @superbobry 

if isHealthy keeps being false for longer than checkInterval, you already jumped from while loop, what's the next....

what you want to do is actually blocking the current thread before you have nWorkers running tasks....what's the problem with the below flow? 

```scala
// Checking Thread
sync.synchronized {
   //checking num of running tasks
   while (numRunningTasks < nWorkers) {
       Thread.sleep(100)
   }
   sync.notify()
}


// Main Thread
sync.synchronized {
   sync.wait(maxLength)
   if (numCores < nWorkers) {
      // stopApplication
   } else {
      // start processing
   }
}


```

Reply:
I think the expected behaviour is:
(1) The checking thread and the training thread should run in parallel.
(2) The checking should throw an Exception to stop the training when numCores go below nWorkers.
Your proposed solution seems only doing checking before the training start, doesn't it?

Reply:
I thought `isHealthy` is the implementation you were expecting. My origin solution would be
```
while (true) {
       Thread.sleep(checkInterval)
       check()
} 
```
I still think it make more sense.

Reply:
`if isHealthy keeps being false for longer than checkInterval, you already jumped from while loop, what's the next....`

and 

how you guarantee that the checking thread can detect numCores goes below nWorkers before the executor comes back

Reply:
> if isHealthy keeps being false for longer than checkInterval, you already jumped from while loop, what's the next....

Let me know if the current implementation makes sense to you.

> how you guarantee that the checking thread can detect numCores goes below nWorkers before the executor comes back

I think we should allow executor to come back and recover the training. That's why I set the default check interval to 30 minutes. If the executor does not come back within 30 minutes, I think the training is unlikely to succeed. Correct me if I miss something?

Reply:
did you test the second item? I think something in Rabit makes it unfeasible for recovery (Scala-based RabitTracker handles it explicitly by stopping the training)

Reply:
kk. Good to know. Will make a test...

Reply:
@CodingCat I think you are right. I experimented by killing one of the executors and the job actually hang forever even if the executor came back. I made some changes to my implementation but this case is still very hard to catch. I would appreciate for any further suggestions.